src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -1174,7 +1174,15 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
       } else tree.tpe match {
         case atp @ AnnotatedType(_, _) if canAdaptAnnotations(tree, this, mode, pt) => // (-1)
           adaptAnnotations(tree, this, mode, pt)
-        case ct @ ConstantType(value) if mode.inNone(TYPEmode | FUNmode) && (ct <:< pt) && canAdaptConstantTypeToLiteral => // (0)
+        case ct @ ConstantType(value) if mode.inNone(TYPEmode | FUNmode | PATTERNmode) && (ct <:< pt) && canAdaptConstantTypeToLiteral => // (0)
+          /* We don't adapt in PATTERNmode because patmat checks for duplicated
```

Comment:
Is this part of the change covered by the test? I could only envisage it mattering in code like `case Foo(1 | 1)`, which we don't want to constant fold based on the pattern's sharper type `lub(1.type, 1.type)` down to `case Foo(1)` before patmat has issued the warning.